flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/exceptionhistory/FailureHandlingResultSnapshot.java

```diff
@@ -0,0 +1,146 @@
+                        String.format(
+                                "The execution %s didn't provide a failure info even though the corresponding ExecutionVertex %s is marked as having handled the root cause of this failure.",
+                                executionVertex.getCurrentExecutionAttempt().getAttemptId(),
+                                executionVertex.getID()));
+                rootCauseExecutionAttemptId =
+                        executionVertex.getCurrentExecutionAttempt().getAttemptId();
+            }
+
+            final Execution execution = executionVertex.getCurrentExecutionAttempt();
+            if (failureHandlingResult.getVerticesToRestart().contains(executionVertex.getID())
+                    && execution.getFailureInfo().isPresent()) {
+                failedExecutions.put(execution.getAttemptId(), execution);
+            }
+        }
+
+        return new FailureHandlingResultSnapshot(
+                rootCauseExecutionAttemptId,
+                failureHandlingResult.getError(),
+                failureHandlingResult.getTimestamp(),
+                failedExecutions);
+    }
+
+    @VisibleForTesting
+    FailureHandlingResultSnapshot(
+            @Nullable ExecutionAttemptID rootCauseExecution,
+            Throwable rootCause,
+            long timestamp,
+            Map<ExecutionAttemptID, Execution> failedExecutions) {
+        this.rootCauseExecution = rootCauseExecution;
+        this.rootCause = rootCause;
+        this.timestamp = timestamp;
+        this.failedExecutions = failedExecutions;
+    }
+
+    /**
+     * Returns the {@link Execution} that handled the root cause for this failure. An empty {@code
+     * Optional} will be returned if it's a global failure.
+     *
+     * @return The {@link Execution} that handled the root cause for this failure.
+     */
+    public Optional<Execution> getRootCauseExecution() {
+        if (rootCauseExecution == null) {
+            return Optional.empty();
+        }
+
+        return Optional.of(failedExecutions.get(rootCauseExecution));
+    }
+
+    /**
+     * The actual failure that is handled.
+     *
+     * @return The {@code Throwable}.
+     */
+    public Throwable getRootCause() {
+        return rootCause;
+    }
+
+    /**
+     * The time the failure occurred.
+     *
+     * @return The time of the failure.
+     */
+    public long getTimestamp() {
+        return timestamp;
+    }
+
+    /**
+     * All {@link Execution Executions} that failed and are planned to be restarted as part of this
+     * failure handling.
+     *
+     * @return The concurrently failed {@code Executions}.
+     */
+    public Iterable<Execution> getConcurrentlyFailedExecution() {
+        return failedExecutions.values().stream()
+                .filter(execution -> !execution.getAttemptId().equals(rootCauseExecution))
+                .collect(Collectors.toSet());
```

Comment:
nit: Not the most efficient way to do it. Alternatively one could store `rootCauseExecution` and other failed `Executions` in separate fields.